dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/validation/hooks/EventCountValidationHook.java

```diff
@@ -0,0 +1,120 @@
+ */
+
+import org.hisp.dhis.common.OrganisationUnitSelectionMode;
+import org.hisp.dhis.program.Program;
+import org.hisp.dhis.program.ProgramInstanceQueryParams;
+import org.hisp.dhis.program.ProgramInstanceService;
+import org.hisp.dhis.trackedentity.TrackedEntityInstance;
+import org.hisp.dhis.tracker.TrackerImportStrategy;
+import org.hisp.dhis.tracker.domain.Event;
+import org.hisp.dhis.tracker.report.TrackerErrorCode;
+import org.hisp.dhis.tracker.report.ValidationErrorReporter;
+import org.hisp.dhis.tracker.validation.TrackerImportValidationContext;
+import org.hisp.dhis.user.User;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+import static org.hisp.dhis.tracker.report.ValidationErrorReporter.newReport;
+
+/**
+ * @author Morten Svanæs <msvanaes@dhis2.org>
+ */
+@Component
+public class EventCountValidationHook
+    extends AbstractTrackerDtoValidationHook
+{
+    @Autowired
+    protected ProgramInstanceService programInstanceService;
+
+    @Override
+    public int getOrder()
+    {
+        return 305;
+    }
+
+    public EventCountValidationHook()
+    {
+        super( Event.class, TrackerImportStrategy.CREATE_AND_UPDATE );
+    }
+
+    @Override
+    public void validateEvent( ValidationErrorReporter reporter, Event event )
+    {
+        TrackerImportValidationContext validationContext = reporter.getValidationContext();
+        User user = validationContext.getBundle().getUser();
+
+        Program program = validationContext.getProgram( event.getProgram() );
+
+        if ( program.isRegistration() )
+        {
+            TrackedEntityInstance tei = validationContext.getTrackedEntityInstance( event.getTrackedEntity() );
+            ProgramInstanceQueryParams params = new ProgramInstanceQueryParams();
+            params.setProgram( program );
+            params.setTrackedEntityInstance( tei );
+            params.setOrganisationUnitMode( OrganisationUnitSelectionMode.ALL );
+            params.setUser( user );
+
+            int count = programInstanceService.countProgramInstances( params );
+
+            if ( count == 0 )
+            {
+                reporter.addError( newReport( TrackerErrorCode.E1037 )
+                    .addArg( tei )
+                    .addArg( program ) );
+            }
+            else if ( count > 1 )
+            {
+                reporter.addError( newReport( TrackerErrorCode.E1038 )
+                    .addArg( tei )
+                    .addArg( program ) );
+            }
+        }
+        else
+        {
+            //TODO: I don't understand the purpose of this here. This could be moved to preheater?
+            // possibly...(Stian-1.4.20)
+            // For isRegistraion=false, there can only exist a single tei and program instance across the entire program.
+            // Both these counts could potentially be preheated, and then just make sure we just have 1 program instance.?
```

Comment:
This class will be reworked by @luciano-fiandesio. We will move the SQL into preheater, and just verify the results here.

Reply:
This class is also a bit strangely named, we are checking ProgramInstances, not ProgramStageInstances. This will also be dealt with later :) We can commit this as-is for now, since we planned the refactoring work already.